File: java/src/main/java/money_problem/domain/ExchangeRate.java
package money_problem.domain;

import io.vavr.control.Either;

import static io.vavr.API.Left;
import static io.vavr.API.Right;

public class ExchangeRate {
    private final double rate;
    private final Currency currency;

    public ExchangeRate(double rate, Currency currency) {
        this.rate = rate;
        this.currency = currency;
    }

    private static boolean isPositive(double rate) {
        return rate > 0;
    }

    public static Either<Error, ExchangeRate> from(double rate, Currency currency) {
        return isPositive(rate)
                ? Right(new ExchangeRate(rate, currency))
                : Left(new Error("Exchange rate should be greater than 0"));
    }

    public Currency getCurrency() {
        return currency;
    }

    public double getRate() {
        return rate;
    }
}

File: java/src/main/java/money_problem/domain/ExchangeRate.kt
package money_problem.domain

import io.vavr.API.Left
import io.vavr.API.Right
import io.vavr.control.Either

class ExchangeRate private constructor(val rate: Double, val currency: Currency) {
    companion object {
        private fun isPositive(rate: Double): Boolean = rate > 0

        @JvmStatic
        fun from(rate: Double, currency: Currency): Either<Error, ExchangeRate> =
            if (isPositive(rate)) Right(ExchangeRate(rate, currency))
            else Left(Error("Exchange rate should be greater than 0"))
    }
}